File: src/main/java/dev/slne/discord/discord/interaction/command/DiscordCommandManager.java
package dev.slne.discord.discord.interaction.command;

/**
 * The type Discord command manager.
 */
public class DiscordCommandManager {

//	private final List<DiscordCommand> commands;
//
//	/**
//	 * Creates a new DiscordCommandManager.
//	 */
//	public DiscordCommandManager() {
//		this.commands = new ArrayList<>();

//		this.commands.add(new TicketButtonCommand());
//		this.commands.add(new TicketCloseCommand());
//		this.commands.add(new TicketMemberAddCommand());
//		this.commands.add(new TicketMemberRemoveCommand());

//		this.commands.add(new TwitchConnectCommand());
//		this.commands.add(new TwitchFollowCommand());
//		this.commands.add(new TicketDependenciesNotMetCommand());

//		this.commands.add(new WhitelistCommand());
//		this.commands.add(new WhitelistedCommand());
//		this.commands.add(new WhitelistQueryCommand());
//		this.commands.add(new WhitelistRoleRemoveCommand());

//		this.commands.add(new ReactionRoleTextCommand());
//	}

//	/**
//	 * Finds a command
//	 *
//	 * @param name The name of the command.
//	 *
//	 * @return The command.
//	 */
//	public DiscordCommand findCommand(String name) {
//		return this.commands.stream().filter(command -> command.getName().equals(name)).findFirst().orElse(null);
//	}

//	/**
//	 * Registers the commands to a guild.
//	 *
//	 * @param guild The guild.
//	 */
//	public void registerToGuild(@NotNull Guild guild) {
//		CommandListUpdateAction updateAction = guild.updateCommands();
//		List<CommandData> commandDatas = new ArrayList<>();
//
//		this.commands.forEach(command -> {
//			SlashCommandData slashCommandData = Commands.slash(command.getName(), command.getDescription());
//			slashCommandData.setGuildOnly(command.isGuildOnly());
//			slashCommandData.setNSFW(command.isNsfw());
//			slashCommandData.addSubcommands(command.getSubCommands());
//			slashCommandData.addOptions(command.getOptions());
//			commandDatas.add(slashCommandData);
//		});
//
////        updateAction.addCommands(commandDatas).queue(
////                cmds -> {
////                    String cmdNames = cmds.stream().map(Command::getName).reduce((a, b) -> a + ", " + b).orElse("");
////                    DataApi.getDataInstance().logInfo(getClass(), String.format("Registered commands [%s] to guild %s.",
////                            cmdNames, guild.getName()));
////                },
////                throwable -> {
////                    DataApi.getDataInstance().logError(getClass(), String.format("Failed to register commands to " +
////                            "guild %s", guild.getName()), throwable);
////                });
//
//		DataApi.getDataInstance()
//			   .logInfo(
//					   getClass(),
//					   String.format("Registered commands [%s] to guild %s.", commandDatas, guild.getName())
//			   );
//	}
}

File: src/main/kotlin/dev/slne/discord/discord/interaction/command/DiscordCommandManager.kt
package dev.slne.discord.discord.interaction.command

import dev.minn.jda.ktx.coroutines.await
import dev.slne.discord.annotation.DiscordCommandMeta
import dev.slne.discord.discord.interaction.command.commands.misc.*
import dev.slne.discord.discord.interaction.command.commands.ticket.*
import dev.slne.discord.discord.interaction.command.commands.ticket.members.TicketMemberAddCommand
import dev.slne.discord.discord.interaction.command.commands.ticket.members.TicketMemberRemoveCommand
import dev.slne.discord.discord.interaction.command.commands.whitelist.*
import it.unimi.dsi.fastutil.objects.Object2ObjectMap
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import it.unimi.dsi.fastutil.objects.ObjectArrayList
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.interactions.commands.build.CommandData
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.kyori.adventure.text.logger.slf4j.ComponentLogger
import kotlin.reflect.full.findAnnotation

object DiscordCommandManager {
    private val logger = ComponentLogger.logger(DiscordCommandManager::class.java)
    private val commands: Object2ObjectMap<String, DiscordCommandHolder> =
        Object2ObjectOpenHashMap()

    init {
        register(WhitelistCommand)
        register(WhitelistedCommand)
        register(WhitelistQueryCommand)
        register(WhitelistRoleRemoveCommand)
        register(WhitelistUnblockCommand)
        register(WhitelistChangeCommand)

        register(TicketButtonCommand)
        register(TicketCloseCommand)
        register(TicketDependenciesNotMetCommand)
        register(TicketFixCommand)
        register(TicketReplyDeadlineCommand)

        register(TicketMemberAddCommand)
        register(TicketMemberRemoveCommand)

        register(DontAskToAsk)
        register(HowToJoin)
        register(FAQCommand)
        register(MissingInformationCommand)

        register(RequestRollbackCommand)
    }

    private fun register(command: DiscordCommand) {
        val annotation = command::class.findAnnotation<DiscordCommandMeta>()
            ?: error("Command $command does not have a DiscordCommandMeta annotation")
        check(annotation.name !in commands) { "Duplicate command handler id ${annotation.name}" }

        val holder = DiscordCommandHolder(annotation, command)
        commands[annotation.name] = holder

        logger.info(
            "Found command ${command.javaClass.simpleName} with name ${annotation.name}"
        )
    }

    fun getCommand(name: String): DiscordCommandHolder? = commands[name]

    suspend fun updateCommands(guild: Guild) {
        logger.info("Starting to update commands for guild ${guild.name} (${guild.id})")

        val commandData = ObjectArrayList<CommandData>(commands.values.size)

        for (commandHolder in commands.values) {
            commandData.add(
                Commands.slash(commandHolder.meta.name, commandHolder.meta.description)
                    .setGuildOnly(commandHolder.meta.guildOnly)
                    .setNSFW(commandHolder.meta.nsfw)
                    .addSubcommands(commandHolder.command.subCommands)
                    .addOptions(commandHolder.command.options)
//                    .setDefaultPermissions(commandHolder.command().getDefaultMemberPermissions()) // TODO: 24.08.2024 10:11 - why not in use?
            )
        }


        guild.updateCommands().await()

        val updatedCommandNames =
            guild.updateCommands().addCommands(commandData).await().map { it.name }

        logger.info(
            "Updated ${updatedCommandNames.size} commands for guild ${guild.name} (${guild.id}) with names: $updatedCommandNames"
        )
    }
}

data class DiscordCommandHolder(val meta: DiscordCommandMeta, val command: DiscordCommand)
